contents/security/passwordStorage.md

```diff
@@ -0,0 +1,98 @@
+
+### Why isn't hashing enough?
+
+Unfortunately, [rainbow tables](https://en.wikipedia.org/wiki/Rainbow_table) exist.
+
+A rainbow table is a precomputed table of hashes for some set of passwords. Basically, since the hashing function is easy to compute one-way, people build huge tables of hashes wherein the plaintext is already known, so that attempting to crack hashes becomes a simple problem of looking up the hash in the table and it's corresponding value, instead of attempting to reverse the hash. Because of this, it is very easy to crack simple hashes by simply doing a lookup.
+
+An example of a service that provides this is [Crackstation](https://crackstation.net/).
+
+Since attackcs like rainbow tables exist, passwords need another layer of security.
+
+## Adding salt
+
+A salt refers to appending a string of text, unique to each user, to their password before hashing them. Since each user has a unique salt, this makes rainbow tables ineffective, as the majority of the precomputed hashes won't even contain the salt, so they wouldn't even matter anyway! In this way, the salt forces the attacker to recompute the rainbow table for each password in order to be able to effectively use it. This effectively converts the attack to brute force, as each hash must be recomputed for each possible password. Additionally, the computed rainbow table would only be useful for that specific user, as each user would have a different salt. It could take years before a password is cracked!
+
+Thus, the way to store passwords properly is to use a salted hash - take the user's password, append some bytes to it, and hash the result. That hash is the user's password. When the user attempts to login again, perform the same procedure again. If the hash matches, then you know that the user is who he says he is, even if you don't actually know the original password that the user provided. 
+
+One question that is commonly asked by developers is where to store the salt. The salt can be stored in plaintext, along with the user in the database. Since the goal of the salt is only to prevent precomputed rainbow tables from being used, it doesn't need to be encrypted in the database.
+
+### If an attacker has gained access to the entire application, what does it matter if my passwords are stored in plaintext or not?
+
+If an attacker has already gained access to the entire application, then he already has all the information hat he coculd possibly get. He would have access to all of your application's data, including data from your users or from any analytics software that you might be running. However, by adding salt and hashing your passwords, the attacker still doesn't know your customer's passwords and could take years to find out. Otherwise, since 59% of people use the same password across multiple sites<sup>1</sup>, the attacker could quickly try other websites such as banks to attempt to break into those accounts, which can yield great financial return.
```

Comment:
Linking to the source for the statistic here might be good 🙂 

Also, do you think that the last line should be phrased to emphasise the detriment to the user, rather than the benefit to the attacker?

Reply:
Is there a way to link it nicely in markbind? The way I did it here was adding a superscript 1 and having the source on the very bottom. I suppose I could always write one up with HTML and CSS.

Reply:
I guess the simplest way would be to just use a markdown hyperlink?

`since 59% of people `**`are reported`**` to use the same password across multiple sites according to [Source](link_to_source)`

I haven't tried it, but you can alternatively embed a hyperlink in the superscript 1 itself.